### PR TITLE
Getting Lottie metadata piped through.

### PR DESCRIPTION
**Pull request by simeoncran**
Adding Lottie metadata to the TranslationResult in GenericData format.
Some improvements of GenericData as a result of using it more.
Adding code to convert from GenericData to Windows.Data.Json. Not yet used but will be added to LottieVisualSource.
Replacing all of our custom empty arrays with Array.Empty<T>().

### CODE REVIEW

File: source/GenericData/GenericDataBool.cs [hunk 1]
```diff
@@ -23,6 +23,8 @@ abstract class GenericDataBool : GenericDataObject
 
         public override GenericDataObjectType Type => GenericDataObjectType.Bool;
 
+        public static implicit operator GenericDataBool(bool value) => Create(value);
```
Comment (scott-moore-ms) [inline source/GenericData/GenericDataBool.cs]: With this existing, would/should anyone ever call the public Create method or should that become private?

File: source/GenericData/GenericDataList.cs [hunk 2]
```diff
@@ -45,5 +46,7 @@ public override string ToString()
             => Count == 0
                 ? "[]"
                 : $"[{string.Join(", ", _items.Select(x => ToString(x)))}]";
+
+        public static implicit operator GenericDataList(GenericDataObject[] value) => Create(value);
```
Comment (scott-moore-ms) [inline source/GenericData/GenericDataList.cs]: >GenericDataObject [](start = 56, length = 17)

Why specifically a GenericDataObject[] vs IEnumerable<GenericDataObject> as the Create() method takes?

File: source/GenericData/GenericDataObject.cs [hunk 3]
```diff
@@ -11,6 +13,14 @@ abstract class GenericDataObject
     {
         public abstract GenericDataObjectType Type { get; }
 
+        public static implicit operator GenericDataObject(bool value) => GenericDataBool.Create(value);
```
Comment (scott-moore-ms) [inline source/GenericData/GenericDataObject.cs]: >GenericDataObject [](start = 40, length = 17)

Oh interesting - why not => (GenericDataBool)value ?

File: source/LottieToWinComp/TranslationResult.cs [hunk 4]
```diff
@@ -19,18 +20,25 @@ sealed class TranslationResult
...
             Visual rootVisual,
             IEnumerable<TranslationIssue> translationIssues,
-            uint minimumRequiredUapVersion)
+            uint minimumRequiredUapVersion,
+            GenericDataMap sourceMetadata)
         {
             RootVisual = rootVisual;
             TranslationIssues = translationIssues.ToArray();
             MinimumRequiredUapVersion = minimumRequiredUapVersion;
+            SourceMetadata = sourceMetadata;
```
Comment (scott-moore-ms) [inline source/LottieToWinComp/TranslationResult.cs]: >SourceMetadata [](start = 12, length = 14)

Should we assert that this is non-null?

File: source/GenericData/GenericDataBool.cs (hunk 1 above)
Comment (simeoncran) [reply source/GenericData/GenericDataBool.cs]: The implicit conversions are handy but I want to allow callers to be explicit if it makes things clearer at the callsite.

---
In reply to: [353405137](https://github.com/windows-toolkit/Lottie-Windows/pull/195#discussion_r353405137) [](ancestors = 353405137)

File: source/GenericData/GenericDataList.cs (hunk 2 above)
Comment (simeoncran) [reply source/GenericData/GenericDataList.cs]: Implicit operators don't accept interfaces as arguments.

---
In reply to: [353405786](https://github.com/windows-toolkit/Lottie-Windows/pull/195#discussion_r353405786) [](ancestors = 353405786)

File: source/GenericData/GenericDataObject.cs (hunk 3 above)
Comment (simeoncran) [reply source/GenericData/GenericDataObject.cs]: Given the choice between an explicit cast that converts into a call to Create, and just calling the method, they both take up the same space but the latter is clearer to me.


---
In reply to: [353407114](https://github.com/windows-toolkit/Lottie-Windows/pull/195#discussion_r353407114) [](ancestors = 353407114)

File: source/LottieToWinComp/TranslationResult.cs (hunk 4 above)
Comment (simeoncran) [reply source/LottieToWinComp/TranslationResult.cs]: Good philosophy question. :-) Once we can switch to C# 8 we can use the type system to make it clear whether this value is optional or not. Let's discuss in person.

---
In reply to: [353412670](https://github.com/windows-toolkit/Lottie-Windows/pull/195#discussion_r353412670) [](ancestors = 353412670)

**Pull request merged**